### PR TITLE
feat: Factory Transfer Voucher

### PR DESCRIPTION
**Pull request by joemonem**
# Pull Request

## Description
This adds the ability to transfer voucher from user to another. 

## Type of Change

- [ ] Bug fix
- [x ] New feature
- [ ] Breaking change
- [ ] Documentation update
- [ ] Refactor/Chore

## Related Issue

[Link to related issue, if applicable]

## Testing

Steps to test:

1. [Step 1]
2. [Step 2]
3. [Step 3]

## Checklist

- [x ] Self-review completed
- [ ] Tests added/updated (if applicable)
- [ ] Documentation updated (if applicable)

## Screenshots (if applicable)

[Add screenshots here]

## Additional Notes

[Any additional information or context]

### CODE REVIEW

File: packages/euclid/src/msgs/factory/msg.rs [hunk 1]
```diff
@@ -66,6 +66,12 @@ pub enum ExecuteMsg {
         cross_chain_addresses: Vec<CrossChainUserWithLimit>,
         timeout: Option<u64>,
     },
+    TransferVirtualBalance {
+        token: Token,
+        amount: Uint128,
+        recipient_address: CrossChainUserWithLimit,
```
Comment (SlayerAnsh) [inline packages/euclid/src/msgs/factory/msg.rs]: CrossChainUser, limit is not needed for transfering Virtual Balance

File: packages/euclid/src/msgs/router.rs [hunk 2]
```diff
@@ -37,6 +37,13 @@ pub enum ExecuteMsg {
         cross_chain_addresses: Vec<CrossChainUserWithLimit>,
         timeout: Option<u64>,
     },
+    TransferVoucher {
```
Comment (SlayerAnsh) [inline packages/euclid/src/msgs/router.rs]: We should name it same as factory message, TransferVirtualBalance

File: contracts/hub/router/src/contract.rs [hunk 3]
```diff
@@ -131,6 +132,42 @@ pub fn execute(
...
+                    &mut deps,
+                    env,
+                    info,
+                    token,
+                    recipient,
+                    amount,
+                    cross_chain_addresses,
+                    timeout,
+                ),
+                ExecuteMsg::TransferVirtualBalanceInternal {
```
Comment (SlayerAnsh) [inline contracts/hub/router/src/contract.rs]: This is probably not needed. A reusable function that can be used above in TransferVoucher and Ibc receive will be sufficient.

File: contracts/hub/router/src/contract.rs [hunk 4]
```diff
@@ -131,6 +132,42 @@ pub fn execute(
                     cross_chain_addresses,
                     timeout,
                 ),
+                ExecuteMsg::TransferVoucher {
+                    token,
+                    recipient,
+                    amount,
+                    cross_chain_addresses,
```
Comment (SlayerAnsh) [inline contracts/hub/router/src/contract.rs]: CrossChainAddresses not needed, all funds should be transfered to recipient

**Review by SlayerAnsh — CHANGES_REQUESTED** [review]

File: packages/euclid/src/msgs/router.rs [hunk 5]
```diff
@@ -37,6 +37,12 @@ pub enum ExecuteMsg {
         cross_chain_addresses: Vec<CrossChainUserWithLimit>,
         timeout: Option<u64>,
     },
+    TransferVirtualBalance {
+        token: Token,
+        recipient: CrossChainUser,
+        amount: Option<Uint128>,
```
Comment (SlayerAnsh) [inline packages/euclid/src/msgs/router.rs]: Amount should be required

File: contracts/hub/router/src/execute.rs [hunk 6]
```diff
@@ -402,6 +402,31 @@ pub fn execute_release_escrow(
...
+    env: Env,
+    _info: MessageInfo,
+    token: Token,
+    recipient: CrossChainUser,
+    amount: Option<Uint128>,
+    timeout: Option<u64>,
+) -> Result<Response, ContractError> {
+    let msg = CosmosMsg::Wasm(WasmMsg::Execute {
+        contract_addr: env.contract.address.to_string(),
+        msg: to_json_binary(&ExecuteMsg::TransferVirtualBalance {
```
Comment (SlayerAnsh) [inline contracts/hub/router/src/execute.rs]: This is calling same function, infinite callback. We want to call virtual balance Transfer Method with sender being info.sender and chain uid is VSL chain uid

File: contracts/hub/router/src/ibc/receive.rs [hunk 7]
```diff
@@ -187,11 +187,34 @@ pub fn reusable_internal_call(
                     tx_id: msg.tx_id,
                 }))?))
         }
+        ChainIbcExecuteMsg::Transfer(msg) => {
+            ibc_execute_transfer_virtual_balance(deps.branch(), env, msg)
```
Comment (SlayerAnsh) [inline contracts/hub/router/src/ibc/receive.rs]: Add a check for sender chain uid and chain uid for secure msgs

File: packages/euclid/src/virtual_balance.rs [hunk 8]
```diff
@@ -37,3 +40,32 @@ impl BalanceKey {
         })
     }
 }
+
+pub fn transfer_virtual_balance(
```
Comment (SlayerAnsh) [inline packages/euclid/src/virtual_balance.rs]: IMO this should be part of contract crate. Also looks very simple so maybe we don't need this function, instead directly use transfer msg? What do you think?

File: contracts/hub/router/src/execute.rs [hunk 9]
```diff
@@ -402,6 +402,31 @@ pub fn execute_release_escrow(
         .add_submessages(release_msgs))
 }
 
+pub fn execute_transfer_voucher_internal(
```
Comment (SlayerAnsh) [inline contracts/hub/router/src/execute.rs]: Seems like we can remove this msg from router. If someone wants to transfer vcoin from vsl, they can directly call virtual_balance transfer method.

**Review by SlayerAnsh — CHANGES_REQUESTED** [review]

File: packages/euclid/src/virtual_balance.rs (hunk 8 above)
Comment (joemonem) [reply packages/euclid/src/virtual_balance.rs]: Agreed

File: contracts/liquidity/factory/src/execute.rs [hunk 10]
```diff
@@ -937,6 +937,59 @@ pub fn execute_withdraw_virtual_balance(
...
+        state.is_native,
+        channel,
+        timeout,
+    )?;
+
+    Ok(Response::new()
+        .add_event(tx_event(
+            &tx_id,
+            info.sender.as_str(),
+            TxType::WithdrawVirtualBalance,
```
Comment (SlayerAnsh) [inline contracts/liquidity/factory/src/execute.rs]: Typo - Wrong tx event attached here

**Review by SlayerAnsh — APPROVED** [review]

**Pull request merged**